### ISSUE TITLE
Relative image paths are incorrectly set

### ISSUE THREAD

**Issue by VictorZakharov** [RickStrahl/MarkdownMonster#619]
So I pasted the image and saved into my images folder, after which MM place this markup

```
![](../../static/images/test.png)
```

Unfortunately, this does not work when deployed. Hugo, the static site generator I'm using, requires images to be placed into "static" folder, but when deployed, it will be available under

```
images/test.png
```

So it would be great if MM used this markup instead:
```
![](/images/test.png)
```

Also a great feature would be using certain format for naming images, for example article/post file name + "_1", 2, 3 etc... then I don't need to spend time figuring a new name for it.

Is it possible to configure any of this?

**Comment by RickStrahl** [comment]
As to image paths there's nothing that can really help with this currently because the document knows nothing about your environment so it doesn't make any assumptions about where things are coming from. So the only option is to fix up the path. If you paste it'll create a relative path two levels down and anything lower will create a physical OS path.

For rendering there `previewWebRootPath` that you can set in the YAML header of the document:

[Fixing up the Preview Root Path with previewWebRootPath](https://markdownmonster.west-wind.com/docs/_5fz0ozkln.htm)

which will can help with rendering your content out of a different path than the current page. 

As to image naming - this is handled at the document level not at the Weblog Addin level. I'm not sure how you would specify that in a way that'd be easy to specify. If you have more specific idea how this needs to work, please open a new issue for this and we can see if this can be added. I think you're asking for this for blog posts in which case this would be tagged to the Weblog Addin - maybe with a custom image dialog specific to blog posts.

MM does remember the last folder you saved, so what I do if I use repetitive names is to select a file with the same name and change the number.

**Comment by VictorZakharov** [comment]
`previewWebRootPath` is great, but is it possible to specify once in settings of the project / MM, to avoid changing front matter for every article? Found a ticket #453, where you added `previewWebRootPath`. Side note - a similar approach is used by typora - adding custom front matter to every article.

One of the benefits of static website generators is to avoid vendor lock-in. I can move my content to another static website generator without much effort. Which I already tried by moving from Jekyll to Hugo, took me <1hr, then some extra to adjust a Hugo style theme to my liking.

Obviously adding any front matter specific to the editor is vendor lock-in, in this case vendor of the editor I'm using. It also means I now need to maintain a script to add/remove vendor specific front matter from my md files, not only when switching editors, but also if path on disk changes.

MM is not free, and paid tools should provide convenience to justify the purchase. My biggest problem is adding images right now. Doesn't take a lot of time, but drains my brain due to many clicks, typing and screen switches (via Greenshot/Paint).

For article naming, the formatter should support at minimum two entities, original article file name without md extension, like {filename} and sequential number {seq} which would be generated dynamically, based on already generated files. And then I could specify {filename}-{seq} as image name format and not have to worry about opening any external windows. Which is very important for focus on writing great content (much like code).

**Comment by VictorZakharov** [comment]
Side question - what's "weblog" I keep seeing in UI, help articles etc? Nothing shows up in google on that. Publish weblog, weblog addin - I feel left out on all those terms.

**Comment by RickStrahl** [comment]
Ok, I made change so that when a project is open the project's path is used as the `PreviewWebRootPath`. IOW, checks go:

* YAML in document
* Project File's Path
* Otherwise not specified and we'll use OS path (likely to break)

As to image editing - please open a new issue so we can tracks these two completely separate issues. You can repaste your above answer there.

Thanks...

**Comment by RickStrahl** [comment]
After a bit of experimenting I decided to add the `PreviewWebRootPath` property to the project. The value uses a specifically set value (in the `.mdproj` JSON file or **Edit Project File**) or if that doesn't exist uses the project files path.

This makes the priority for / path evaluation:

* YAML
* Project File Setting
* Project File Path

Also updated the document for this corresponding change:
http://markdownmonster.west-wind.com/docs/_5fz0ozkln.htm

Changes are here:
1784d1109d1bb0f0a7bc46a6219f1dfd3d3e05a1

**Comment by VictorZakharov** [comment]
Thank you @RickStrahl - this sounds great, I will check it out when it's released.

**Comment by RickStrahl** [comment]
It's in the current downloadable builds (1.20.15.6+)

**Comment by VictorZakharov** [comment]
Oh alright, I looked at release notes and figured you are waiting on more changes to produce a build.
https://github.com/RickStrahl/MarkdownMonster/blob/master/Changelog.md
![image](https://user-images.githubusercontent.com/8978766/72929573-18a9f880-3d28-11ea-860f-b25cc9044ad4.png)

**Comment by VictorZakharov** [comment]
@RickStrahl  Installed latest version, this feature doesn't work for me.

What works (image is displayed):
![svn error guid forbidden](C:/Git/neolisk-blog-hugo/static/images/2013-02-14-hard-disk-sentinel-fix-disk-perf-mon-1.png)

But this doesn't work (image is not displayed):
![svn error guid forbidden](/images/2013-02-14-hard-disk-sentinel-fix-disk-perf-mon-1.png)

In my project file which sits in C:\\Git\\neolisk-blog-hugo I have the following configuration:
"PreviewWebRootPath": "C:\\Git\\neolisk-blog-hugo\\static\\", with or without the trailing backslash - doesn't matter.

Do you know what could be wrong?

**Comment by RickStrahl** [comment]
Try removing the trailing slash?

Take a look at the output (Use Shift-F12 to open in your default browser, then View Source) to see what the actual URL generated is for the links.  The double slash maybe what's breaking the url for the image.

**Comment by VictorZakharov** [comment]
@RickStrahl 
![image](https://user-images.githubusercontent.com/8978766/72991836-b6e99d00-3dc0-11ea-8250-cd71bb857cc9.png)
I did try reopening MM a couple times - did not help.
Looks like the setting was ignored, it just used the path I specified.
Document was loaded from:
```
file:///C:/Users/Victor/AppData/Local/Temp/_MarkdownMonster_Preview.html
```

**Comment by VictorZakharov** [comment]
I tried with or without the trailing slash, also with single slashes vs double - makes no difference.

**Comment by RickStrahl** [comment]
Make sure the project file is opened in MM.

**Comment by VictorZakharov** [comment]
Yes, I double clicked the project file.

**Comment by RickStrahl** [comment]
I'm not sure then. This works for me:

Without a project open:

![image](https://user-images.githubusercontent.com/1374013/73152882-a0b83700-4075-11ea-8f28-c9690613e8c3.png)


With a project open and `PreviewWebRootPath` set:

![image](https://user-images.githubusercontent.com/1374013/73152834-80887800-4075-11ea-8cb6-bbfd1d9df994.png)

**Comment by VictorZakharov** [comment]
@RickStrahl You forgot to include contents of `mdproj` file.

Side note - I found it's possible to do via VS code and plugins. Will be using that for now. I'll be missing a few features from MM (such as table editing via UI), but this image workflow is very important for me.

**Comment by RickStrahl** [comment]
The file just has the PreviewWebRootPath in it:

```json
{
  "ActiveFolder": "c:\\projects\\MarkdownMonster",
  "ActiveSidebarIndex": 0,
  "OpenDocuments": [
    {
      "Filename": "c:\\projects\\MarkdownMonster\\README.md",
      "IsActive": false,
      "LastEditorLineNumber": 0,
      "LastImageFolder": null
    },
    {
      "Filename": "c:\\projects\\MarkdownMonster\\Changelog.md",
      "IsActive": true,
      "LastEditorLineNumber": 6,
      "LastImageFolder": null
    }
   
  ],
  "PreviewWebRootPath": "C:\\projects\\MarkdownMonster"
}
```

When the project is created or the WebPreviewPath doesn't exist that path will default to the project's path.

Still not sure why that wouldn't be working for you with the project open and the path set correctly.

**Comment by RickStrahl** [comment]
So gave this a little more thought and decided to add a couple more options to make this a little easier. So rather than requiring a separate project that has to be open you can now do one of the following:

* Add a `.markdownmonster` file to your 'project root'
* If you do have a `.mdproj` file it will find that even if project is not open
* Also finds `_toc.json` which is a KavaDocs project file

It's basically the same behavior as before, but with a few more options and most notably not requiring that a project is open.

**Comment by VictorZakharov** [comment]
@RickStrahl  What should be the contents of .markdownmonster ? I'm asking because root is not everything. The actual images are under /static/images but they appear under /images when deployed.

**Comment by RickStrahl** [comment]
For right now, it's just a marker file to find the root.

**Comment by RickStrahl** [comment]
As to those other settings - I think those could go into a project file, but not into the simple marker files because these need to be looked for frequently and that will end up impacting  rendering performance. Projects are open once loaded so the data is always available.

Currently MM handles image paths by remembering the last place you saved an image for a given document and then reopening the image save dialog there.

That combined with OS history in save dialogs that show previously saved files, I personally think address most of the use cases for any file locations without having to configure 20 different file save locations. While there might be a handful of people who want to control every single possible file save option, the vast majority will never find an option like this or even knowing it exists bother to use it.

**Comment by VictorZakharov** [comment]
I'm not sure what you mean by "20 different file save locations". One project - one save location. That's the setup I got, the config is provided above. It's the same config for all .md articles on the website.

**Comment by VictorZakharov** [comment]
I was able to get it more or less working by creating a folder symlink images pointing to static/images. Inserted path is still incorrect it says ../../static/images (content in /content/posts), but I can fix it by removing ../../static. images symlink is added to gitignore. I bought the license, and looking forward to writing some quality content for my blog.

P.S. I gave up on Visual Code Code workflow due to integration complexity of various pieces created by unrelated vendors. Also it had a command line like interface for inserting images (via plugin), you never know if it's working.

@RickStrahl  Thanks.

**Comment by RickStrahl** [comment]
If you really wanted to customize this to just the way you want to do this, you can probably do it with an addin. As I mentioned the problem with this is that people want so many different types of customization as to make it nearly impossible to make anybody happy.

As to image paths. MM remembers the last save location on a per document basis on any images that are open or that are in the recent document list (eventually though they fall off). The image folder location remembering is:

* Last folder used for images (per document)
* Last image folder used (if above is not available/per session)
* Document folder

I think in your case you'd end up having to set the image path only the first time and then you're in the right place.

As for image linking if you use absolute base paths (ie. `/images/blah.png`) MM will resolve the `/`  based on the various rules including the `.markdownmonster` or `.mdproj` file you point out. But I assume you're already doing that.

I haven't forgotten about this - everytime I use SnagIt and see the confusing image naming template I'm reminded :smile:

**Comment by VictorZakharov** [comment]
It would take me a few days to figure out how to write an addin, probably not worth the effort.

I think there should be an object array as configuration, each item would have path on disk and hosted path. People can set as many as they want. VirtualFolders or smth like that. I only need one object in there, for images. For files it does not matter if path is correct or not. I'm okay with a preview for files. But I can see when people might need several config objects if they manage multiple image collections on their blog (not just posts).

Absolute paths won't help, because it picks images from "static/images", but it doesn't know I have a symbolic link "images", so it doesn't know to optimize the path to it. Relative or absolute, I would have to manually fix it. Might be slightly less typing with absolute paths.

What's more annoying though, is that renaming a file does not honor its git status. So MM just removes the file and creates another one - this results in broken git commit (missing file). Instead, it should property rename as Visual Studio does (or IntelliJ with their rename refactor), keeping all git history for that file.

Should I create a separate issue for this? Looks like a generic issue for which there can be a universal fix that would work for everyone, not just my use case.

@RickStrahl
